File: services/workflow-automation-service/src/models/execution.rs
use chrono::{DateTime, Utc};
use event_bus::contracts::WorkflowTriggerRequested;
use serde::{Deserialize, Serialize};
use serde_json::Value;
use sqlx::FromRow;
use uuid::Uuid;

#[derive(Debug, Clone, FromRow, Serialize, Deserialize)]
pub struct WorkflowRun {
    pub id: Uuid,
    pub workflow_id: Uuid,
    pub trigger_type: String,
    pub status: String,
    pub started_by: Option<Uuid>,
    pub current_step_id: Option<String>,
    pub context: Value,
    pub error_message: Option<String>,
    pub started_at: DateTime<Utc>,
    pub finished_at: Option<DateTime<Utc>>,
}

#[derive(Debug, Deserialize)]
pub struct StartRunRequest {
    #[serde(default)]
    pub context: Value,
}

#[derive(Debug, Deserialize)]
pub struct TriggerEventRequest {
    #[serde(default)]
    pub context: Value,
}

#[derive(Debug, Deserialize, Default)]
pub struct InternalLineageRunRequest {
    #[serde(default)]
    pub context: Value,
}

pub type InternalTriggeredRunRequest = WorkflowTriggerRequested;

#[derive(Debug, Deserialize)]
pub struct ListRunsQuery {
    pub page: Option<i64>,
    pub per_page: Option<i64>,
}

File: services/workflow-automation-service/internal/models/execution.go
package models

import (
	"encoding/json"
	"time"

	"github.com/google/uuid"
)

// WorkflowRun mirrors models::execution::WorkflowRun.
type WorkflowRun struct {
	ID            uuid.UUID       `json:"id"`
	WorkflowID    uuid.UUID       `json:"workflow_id"`
	TriggerType   string          `json:"trigger_type"`
	Status        string          `json:"status"`
	StartedBy     *uuid.UUID      `json:"started_by"`
	CurrentStepID *string         `json:"current_step_id"`
	Context       json.RawMessage `json:"context"`
	ErrorMessage  *string         `json:"error_message"`
	StartedAt     time.Time       `json:"started_at"`
	FinishedAt    *time.Time      `json:"finished_at"`
}

// StartRunRequest mirrors models::execution::StartRunRequest.
type StartRunRequest struct {
	Context json.RawMessage `json:"context,omitempty"`
}

// TriggerEventRequest mirrors models::execution::TriggerEventRequest.
type TriggerEventRequest struct {
	Context json.RawMessage `json:"context,omitempty"`
}

// InternalLineageRunRequest mirrors models::execution::InternalLineageRunRequest.
type InternalLineageRunRequest struct {
	Context json.RawMessage `json:"context,omitempty"`
}

// InternalTriggeredRunRequest aliases the WorkflowTriggerRequested
// payload from libs/event-bus-control. Mirrors the Rust type alias
// `pub type InternalTriggeredRunRequest = WorkflowTriggerRequested`.
type InternalTriggeredRunRequest struct {
	WorkflowID    uuid.UUID       `json:"workflow_id"`
	TriggerType   string          `json:"trigger_type"`
	StartedBy     *uuid.UUID      `json:"started_by,omitempty"`
	Context       json.RawMessage `json:"context,omitempty"`
	CorrelationID uuid.UUID       `json:"correlation_id,omitempty"`
}

// ListRunsQuery mirrors models::execution::ListRunsQuery.
type ListRunsQuery struct {
	Page    *int64
	PerPage *int64
}
